src/python/pants/engine/target.py

```diff
@@ -1779,21 +1782,43 @@ def file_path(self) -> str | None:
         This works without hydration because we validate that `*` globs and `!` ignores are not
         used. However, consider still hydrating so that you verify the source file actually exists.
 
-        The return type is optional because it's possible to have 0-1 files. Most subclasses
-        will have 1 file, though.
+        The return type is optional because it's possible to have 0-1 files.
         """
         if self.value is None:
             return None
         return os.path.join(self.address.spec_path, self.value)
 
     @property
     def globs(self) -> tuple[str, ...]:
-        # Subclasses might override `required = False`, so `self.value` could be `None`.
         if self.value is None:
             return ()
         return (self.value,)
 
 
+class SingleSourceField(OptionalSingleSourceField):
+    """The `source: str` field.
+
+    Unlike `OptionalSingleSourceField`, the `.value` must be defined, whether by setting the
+    `default` or making the field `required`.
+
+    See the docstring for `SourcesField` for some class properties you can set, such as
+    `expected_file_extensions`.
+
+    When you need to get the sources for all targets, use `tgt.get(SourcesField)` rather than
+    `tgt.get(SingleSourceField)`.
+    """
+
+    required = True
+    expected_num_files = 1
+    value: str
+
+    @property
+    def file_path(self) -> str:
+        result = super().file_path
+        assert result is not None
+        return result
```

Comment:
By having `value: str` don't you need a `compute_value() -> str` with matching signature..? Rather than the optional str that's inherited..

Reply:
No, the type hints aren't couples with MyPy.

Reply:
Cool 😎